### PR TITLE
ci(sdk): Autogenerate changelog on build

### LINKED ISSUE #42750: Implement changelog generation as a part of SDK release process
This requires https://github.com/metabase/metabase/pull/42516 to be merged first.

We can hook this up as a part of the PR (for version bump) creation process for simplicity.

### PR DESCRIPTION
**Pull request by WiNloSt**
Closes #42750

The successful CI run [can be seen here.](https://github.com/metabase/metabase/actions/runs/9317795703)

Here's the opened PR to update[ package.json template and changelog](https://github.com/metabase/metabase/pull/43440). Please note that there's no readme update anymore since we removed the specific version from is and refer to 50 RC instead.

### CODE REVIEW

File: .github/workflows/release-embedding-sdk.yml [hunk 1]
```diff
@@ -186,8 +183,43 @@ jobs:
...
+        with:
+          name: final-metabase-sdk
+          path: ./resources/embedding-sdk
+
       - name: Publish to NPM
         run: |
           cd sdk
           echo //registry.npmjs.org/:_authToken=${{ secrets.NPM_RELEASE_TOKEN }} > .npmrc
-          npm publish
+          npm publish --dry-run
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:199]: This ensure we don't publish anything until we're sure this workflow runs like what we expected.

File: .github/workflows/release-embedding-sdk.yml [hunk 2]
```diff
@@ -30,25 +30,21 @@ jobs:
         uses: actions/checkout@v4
         with:
           ref: ${{ inputs.git_ref }}
+          fetch-depth: 0
+          fetch-tags: true
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:34]: 1. `fetch-tags: true` [doesn't work without](https://github.com/actions/checkout/issues/1471#issuecomment-1771231294) `fetch-depth: 0` and this looks like a bug.

2. I need this change because otherwise `git tag -l` wouldn't return anything because GitHub action does shallow fetch by default.

File: .github/workflows/release-embedding-sdk.yml [hunk 3]
```diff
@@ -30,25 +30,21 @@ jobs:
...
 
-      - name: Setup git user
-        run: |
-          git config --global user.email "metabase-bot@metabase.com"
-          git config --global user.name "Metabase bot"
-
-      - name: Create a new git tag
+      - name: check tags
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:36]: I moved the tag step later as [mentioned here.](https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-cli#recommended-workflow)

> The reason why you should commit and tag after `conventionalChangelog` is that the CHANGELOG should be included in the new release, hence `gitRawCommitsOpts.from` defaults to the latest semver tag.

tl;dr We need to tag after generating the changelog otherwise, we won't get the correct changelog.

File: .github/workflows/release-embedding-sdk.yml [hunk 4]
```diff
@@ -30,25 +30,21 @@ jobs:
...
-
-      - name: Push the new tag
-        id : push-tag
-        run: |
-          git push origin ${{ env.tag }}
-      - if: ${{ failure() && steps.push-tag.outcome == 'failure' }}
-        run: echo "Tag '${{ env.tag }}' already exists. If you expect to run this workflow with the same tag, please remove the tag first and rerun this workflow again."
+          if [ $(git tag -l | grep $tag) ]
+          then
+            echo "::error::Tag '${{ env.tag }}' already exists. If you expect to run this workflow with the same tag, please remove the tag first and rerun this workflow again."
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:40]: The message is the same but uses [the syntax](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message) that would surface the error to the job summary making it easier to spot this error.

[Example](https://github.com/metabase/metabase/actions/runs/9302161691)
image

File: .github/workflows/release-embedding-sdk.yml [hunk 5]
```diff
@@ -77,6 +73,8 @@ jobs:
         uses: actions/checkout@v4
         with:
           ref: ${{ inputs.git_ref }}
+          fetch-depth: 0
+          fetch-tags: true
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:77]: We need access to all tags for changelog generation

File: .github/workflows/release-embedding-sdk.yml [hunk 6]
```diff
@@ -86,9 +84,23 @@ jobs:
         with:
           m2-cache-key: "release-sdk"
 
+      - name: Update readme
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:87]: This extra step is crucial because we're now not putting the most updated changelog in the SDK. You can see our [NPM package page](https://www.npmjs.com/package/@metabase/embedding-sdk-react?activeTab=readme) that the version 0.1.5 still refers to 0.1.4 in the readme

You can verify that this is fixed by downloading `metabase-sdk` artifact [from this build](https://github.com/metabase/metabase/actions/runs/9317795703). And check the readme file that it should refer to the correct version. Although it's not that obvious at this moment because we just change all the version reference in SDK readme to 50 RC. So now the SDK readme won't be updated after every SDK release anymore.

File: .github/workflows/release-embedding-sdk.yml [hunk 7]
```diff
@@ -86,9 +84,23 @@ jobs:
...
+        run: |
+          ./bin/embedding-sdk/release_utils.bash update_readme ${{ inputs.sdk_version }}
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:89]: I move the command to a shared script because I'll run it more than once.

File: .github/workflows/release-embedding-sdk.yml [hunk 8]
```diff
@@ -86,9 +84,23 @@ jobs:
...
+
       - name: Bump published npm package version
         run: |
-          sed -i -E 's/("version": ").*"/\1${{ inputs.sdk_version }}"/' enterprise/frontend/src/embedding-sdk/package.template.json
+          ./bin/embedding-sdk/release_utils.bash update_package_json_template ${{ inputs.sdk_version }}
+
+      - name: Upload release utils script, so it can be used later on when checking out `master` branch
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:96]: we could remove this step after the script is merged to master, but for now, let's copy this file over, because we'll need it again when checking out `master` later on.

File: .github/workflows/release-embedding-sdk.yml [hunk 9]
```diff
@@ -86,9 +84,23 @@ jobs:
...
+        uses: actions/upload-artifact@v4
+        with:
+          name: release-util
+          path: ./bin/embedding-sdk/release_utils.bash
+
+      - name: Update changelog
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:102]: I do this now for 2 reasons
1. We should generate the change log with the specified commit. Not when checking out `master` in the later job.
2. This way we ensure CHANGELOG.md is included in the npm package.

File: .github/workflows/release-embedding-sdk.yml [hunk 10]
```diff
@@ -102,6 +114,12 @@ jobs:
           name: metabase-sdk
           path: ./resources/embedding-sdk
 
+      - name: Upload changelog
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:118]: We'll use this when checking out `master` later when we're about to create a PR that update `master` with the latest changelog.

File: .github/workflows/release-embedding-sdk.yml [hunk 11]
```diff
@@ -153,22 +170,40 @@ jobs:
           git config --global user.email "metabase-bot@metabase.com"
           git config --global user.name "Metabase bot"
 
+      - name: Download release utils script
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:175]: We can remove this step once this script is merged to master

File: .github/workflows/release-embedding-sdk.yml [hunk 12]
```diff
@@ -177,6 +211,7 @@ jobs:
                        --assignee "${GITHUB_ACTOR}"\
                        --title "Update SDK version to ${{ inputs.sdk_version }}"\
                        --body "Update Readme version references and published npm package version to ${{ inputs.sdk_version }}"
+          gh pr edit --add-reviewer @metabase/embedding
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:214]: Auto tag our team as a reviewer so we don't forget to merge the PR 😉

File: .github/workflows/release-embedding-sdk.yml [hunk 13]
```diff
@@ -191,3 +226,29 @@ jobs:
           cd sdk
           echo //registry.npmjs.org/:_authToken=${{ secrets.NPM_RELEASE_TOKEN }} > .npmrc
           npm publish
+
+  git-tag:
```
Comment (WiNloSt) [inline .github/workflows/release-embedding-sdk.yml:230]: Moved the git tag step last following the [recommended workflow.](https://github.com/metabase/metabase/issues/42749#issuecomment-2137767536)

File: bin/embedding-sdk/release_utils.bash [hunk 14]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+function update_readme() {
```
Comment (WiNloSt) [inline bin/embedding-sdk/release_utils.bash:2]: These functions are the same as when they were inlined in the SDK release workflow.

File: bin/embedding-sdk/release_utils.bash [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+  sed -i -E 's|(embedding-sdk-)[0-9.]+|\1'"$1"'|' enterprise/frontend/src/embedding-sdk/README.md
+  sed -i -E 's|(http://downloads.metabase.com/sdk/v)[0-9.]+|\1'"$1"'|' enterprise/frontend/src/embedding-sdk/README.md
+}
+
+function update_package_json_template() {
+  sed -i -E 's/("version": ").*"/\1'"$1"'"/' enterprise/frontend/src/embedding-sdk/package.template.json
+}
+
+$1 $2
```
Comment (WiNloSt) [inline bin/embedding-sdk/release_utils.bash:11]: $1 calls the first argument which is the function name e.g.
```bash
release_utils.bash update_readme
```

$2 pass the argument to the function e.g.
```bash
release_utils.bash update_readme 0.1.6
```

`$1 $2` will expand to `update_readme 0.1.6`

And the passed argument could be accessed with $1 (first argument) in each function.

File: enterprise/frontend/src/embedding-sdk/conventional-changelog-config.js [hunk 16]
```diff
@@ -0,0 +1,4 @@
+// eslint-disable-next-line import/no-commonjs
+module.exports = {
+  gitRawCommitsOpts: { grep: "(sdk)" },
```
Comment (WiNloSt) [inline enterprise/frontend/src/embedding-sdk/conventional-changelog-config.js:3]: This is the way to filter the only scope we want. [See more.](https://github.com/metabase/metabase/issues/42749#issuecomment-2137767536)

**Review by npretto — APPROVED** [review]
